STD support for the ESP-IDF framework

ivmarkov opened the pull request:
Dear all,

This PR is implementing libStd support for the [ESP-IDF](https://github.com/espressif/esp-idf) newlib-based framework, which is the open source SDK provided by Espressif for their MCU family (esp32, esp32s2, esp32c3 and all other forthcoming ones).

Note that this PR has a [sibling PR](https://github.com/rust-lang/libc/pull/2310) against the libc crate, which implements proper declarations for all ESP-IDF APIs which are necessary for libStd support.

# Implementation approach

The ESP-IDF framework - despite being bare metal - offers a relatively complete POSIX API based on newlib. `pthread`, BSD sockets, file descriptors, and even a small file-system VFS layer. Perhaps the only significant exception is the lack of support for processes, which is to be expected of course on bare metal.

Therefore, the libStd support is implemented as a set of (hopefully small) changes to the `sys/unix` family of modules, in the form of conditional-compilation branches based either on `target_os = "espidf"` or in a couple of cases - based on `target_env = "newlib"` (the latter was already there actually and is not part of this patch).

The PR also contains two new targets:
- `riscv32imc-esp-espidf`
- `riscv32imac-esp-espidf`

... which are essentially copies of `riscv32imc-unknown-none-elf` and `riscv32imac-unknown-none-elf`, but enriched with proper `linker`, `linker_flavor`, `families`, `os`, `env` etc. specifications so that (a) the proper conditional compilation branches in libStd are selected when compiling with these targets and (b) the correct linker is used.

Since support for atomics is a precondition for libStd, the `riscv32imc-esp-espidf` target additionally is configured in such a way, so as to emit libcalls to the `__sync*` & `__atomic*` GCC functions, which are already implemented in the ESP-IDF framework. If this modification is not acceptable, we can also live with only the `riscv32imac-esp-espidf` target as well.  While the RiscV chips of Espressif lack native atomics support, the relevant instructions are transparently emulated in the ESP-IDF framework using invalid instruction trap. This modification was implemented specifically with Rust support in mind.

# Target maintainers

In case this PR eventually gets merged, you can list myself as a Target Maintainer.

More importantly, Espressif (the chip vendor) is now actively involved and [embracing](https://github.com/espressif/rust-esp32-example/blob/main/docs/rust-on-xtensa.md) all [Rust-related efforts](https://github.com/esp-rs) which were originally a community effort. In light of that, I suppose @MabezDev - who initiated the Rust-on-Espressif efforts back in time and who now works for Espressif won't object to being listed as a maintainer as well.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @dtolnay (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-highfive commented:
:warning: **Warning** :warning:

* These commits modify **compiler targets**. (See the [Target Tier Policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html).)

dtolnay commented:
r? @rust-lang/libs

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3209630114) [(plain)](https://github.com/rust-lang/rust/commit/0305f4afa9757ef3c4764bdfed4f625c1eea563c/checks/3209630114/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 625 error codes
* highest error code: E0783
tidy error: /checkout/library/std/src/os/espidf/mod.rs:3: TODO is deprecated; use FIXME
Found 0 error codes with no tests
Done!
* 345 features
some tidy checks failed
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:13
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3209681861) [(plain)](https://github.com/rust-lang/rust/commit/8d638671a977df9e0aff76a66d5ead28879de2dd/checks/3209681861/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/sys/unix/process/process_unsupported.rs at line 5:
 use crate::num::NonZeroI32;
 use crate::os::raw::NonZero_c_int;
 use crate::sys;
-use crate::sys::unix::unsupported::*;
 use crate::sys::cvt;
 use crate::sys::pipe::AnonPipe;
 use crate::sys::process::process_common::*;
Diff in /checkout/library/std/src/sys/unix/process/process_unsupported.rs at line 12:
+use crate::sys::unix::unsupported::*;
 
 use libc::{c_int, pid_t};
Diff in /checkout/library/std/src/sys/unix/args.rs at line 252:
     use super::Args;
 
 
     #[inline(always)]
-    pub unsafe fn init(_argc: isize, _argv: *const *const u8) {
-    }
+    pub unsafe fn init(_argc: isize, _argv: *const *const u8) {}
     pub fn args() -> Args {
     pub fn args() -> Args {
         Args { iter: Vec::new().into_iter() }
Diff in /checkout/library/std/src/sys/unix/mod.rs at line 46:
 pub use crate::sys_common::os_str_bytes as os_str;
 
 #[cfg(all(not(test), target_os = "espidf"))]
-pub fn init(argc: isize, argv: *const *const u8) {
-}
+pub fn init(argc: isize, argv: *const *const u8) {}
 
 #[cfg(all(not(test), not(target_os = "espidf")))]
 // SAFETY: must be called only once during runtime initialization.
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/unix/process/process_vxworks.rs" "/checkout/library/std/src/sys/unix/args.rs" "/checkout/library/std/src/sys/unix/l4re.rs" "/checkout/library/std/src/sys/unix/process/process_fuchsia.rs" "/checkout/library/std/src/sys/unix/fd.rs" "/checkout/library/std/src/sys/unix/net.rs" "/checkout/library/std/src/sys/unix/weak.rs" "/checkout/library/std/src/sys/unix/mutex.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3209757405) [(plain)](https://github.com/rust-lang/rust/commit/d8e5fb18c4ab19c21c261f234c38df64fcd5e5eb/checks/3209757405/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({"library/std"}) not skipped for "bootstrap::test::Crate" -- not in ["src/tools/tidy"]
 finished in 2.511 seconds
Testing std stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling std v0.0.0 (/checkout/library/std)
error[E0425]: cannot find function `init` in module `sys`
  --> library/std/src/sys_common/rt.rs:16:14
   |
16 |         sys::init(argc, argv);
   |              ^^^^ not found in `sys`
help: consider importing one of these items
   |
4  | use crate::sys::args::init;
   |
   |
4  | use crate::sys::net::init;
   |
4  | use crate::sys::stack_overflow::init;
   |
4  | use crate::sys::thread::guard::init;

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
For more information about this error, try `rustc --explain E0425`.
error: could not compile `std`

To learn more, run the command again with --verbose.


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "std" "--" "--quiet"


Build completed unsuccessfully in 0:19:25
```

</details>

the8472 commented on library/std/src/sys/unix/fd.rs:
```diff
@@ -66,6 +69,10 @@ const fn max_iov() -> usize {
     16 // The minimum value required by POSIX.
 }
 
+// ESP-IDF does not (yet) support duplication of file descriptors, so we are emulating it here
```
That kind of emulation seems dicy since duped file descriptors normally don't share a seek position, nonblocking status and various other properties. Here they would.

How much would break if `duplicate` always returned an error?

ivmarkov replied:
This is here just to support off-the-shelf Rust crates (I've so far only found [Rocket](https://rocket.rs/) relying on it, but I'm unsure how many more are there). But no pressing use case, and Rocket is honestly a bit too big for this MCU anyway.

In any case and based on your explanations, this implementation is as flawed as it gets, so I'll just eradicate it completely tomorrow.

If proper file duplication semantics turns out to be necessary, I don't see any other way besides implementing it in the ESP-IDF framework itself.

ivmarkov replied:
Hey I was [refreshing my knowledge](https://www.gnu.org/software/libc/manual/html_node/Duplicating-Descriptors.html) on `dup()`/`dup2()`/`FD_DUPFD` this morning (as I wrote this code 9 months ago!), and I don't think your interpretation "since duped file descriptors normally don't share a seek position, nonblocking status and various other properties" is correct.

According to the above link, it is exactly the opposite: dup-ed file descriptors DO share seek position, nonblocking status and various other properties. Duped file descriptors are nothing else but aliases of each other, and their only (limited) use seems to be lifetime management, where you can close the original file descriptor while the duped one continues to live and vice versa.

I think more or less this is what my implementation also does, doesn't it? (we can forget about the `CLOEXEC` semantics as that's irrelevant when you don't have processes)?

As for the use case with Rocket (remembered that too) - Rocket was `dup`-ing the file descriptor (socket really) so that it spins a separate socket writer from socket reader, and manages their lifetimes separately too.

the8472 replied:
Yeah, you're right. I guess I was thinking of the mechanism that reopens a file via procfs, which isn't relevant here.

Ignore everything I have said here. 😔

the8472 commented on library/std/src/os/espidf/fs.rs:
```diff
@@ -0,0 +1,126 @@
+    fn st_atime_nsec(&self) -> i64 {
+        //self.as_inner().as_inner().st_atime_nsec as i64
+        0
+    }
```
Since this trait is platform-specific it doesn't need to have methods it can't implement properly.

ivmarkov replied:
Agreed - will remove tomorrow as well. Ditto for all the others there returning 0.

ivmarkov replied:
Btw, would you want me to continue force-pushing, or shall I start accumulating changes-after-feedback as separate commits, so that the progress is easier to track?

the8472 replied:
either is fine. in the latter case you can rewrite the history before merge.

ivmarkov replied:
OK. While we are at the `os/espidf/mod.rs` topic. Any hits regarding what shall I put as a "version" there?
I'm referring to this [FIXME](https://github.com/rust-lang/rust/pull/87666/files#diff-5b9c55eb49b42c03102371ec9b45d87ead04588d68415c07a1166186438a5eb9).

the8472 replied:
Current nightly is 1.56, so if it were to land now and you wanted the new platform features to be instantly stable with the the first release you can put that version there.
If those traits are not critical for the platform to be usable and you have some reason to iterate on them you could mark them unstable instead.

the8472 replied:
Oh wait, that's for the raw types, not the metadata extensions? I guess those are important and should be stable with the the first release.

ivmarkov replied:
Neither is top priority, as using filesystems on the MCU is a bit of a niche case I think, due to the limited flash size in the MCU itself and the need to jump to external SD card for these use cases.

Let me sleep over it to figure out what is most appropriate regarding stable/unstable.
 
Threads and sockets is most important to get right first.

ivmarkov replied:
OK, so I've force-pushed a cleanup of the `os/espidf` module:
* First of all, I decided to leave as stable everything, because, well, filesystems on the ESP-IDF might be a corner case but they are there, the `stat()` function works (even if a bit slow on the SPIFFS FS), so why shall I be seeking for unstable, where it does not make sense
* I've removed all commented out code that should not be there, double-checked that the `raw::stat` module is a faithful representation of what is in there in the ESP-IDF in terms of stat structure & members
* Some stubbed out code *MUST* stay and return 0. Turns out, the "platform specific" `MetadataExt` trait is not so platform specific, as it is used by `os::unix::fs` which is re-exposing its methods to the outside world. Since the ESP-IDF claims to be in the unix family, it *must* support all methods `os::unix::fs` re-expose, even if a subset of those (the `*_nsec` ones) return 0. I'm not setting a precedent: our sibling bare-metal RTOS `vxworks` port [does the same](https://github.com/rust-lang/rust/blob/master/library/std/src/os/vxworks/fs.rs#L82) with `*_nsec`, for the very same reasons.  
* Finally, I've kept versions at 1.1.0 (for the deprecated stuff) and at 1.8.0 for the `MetadataExt` trait as everybody else does. Again, an example: the vxworks port [was merged](https://github.com/rust-lang/rust/pull/61946) ~ 2019, while 1.8.0 is at least 3 years older (2016), let alone version 1.1.0. 
* Moreover: I would actually argue that re-using these versions as they are is even the correct behavior: imagine a cross-platform crate that uses this stuff, either via `os::unix::fs` or `os::platform` and sets the minimum compiler version requirement to 1.8.0. From the point of view of the crate that's the correct behavior, as everybody else except the ESP-IDF port is having those since 1.8.0. Now, people willing to consume that fictitious crate with ESP-IDF will of course never use a 1.8.0 version of the compiler simply because ESP-IDF support is not there. So putting a higher version on those crate items does not seem to help. Quite the opposite.

ivmarkov replied:
@the8472 Just curious if you consider these changes acceptable?

the8472 replied:
Sounds reasonable considering the precedent and that `os::unix` reexports stuff.

the8472 commented:
Note that as part of only supporting fallible allocation in the linux kernel (https://github.com/rust-lang/rfcs/pull/3140) there is ongoing discussion around making std features optional and how to support that on the cargo side. https://github.com/rust-lang/rfcs/pull/3146

I'm not sure what exactly the roadmap on those parts are but it sounds like it would be appropriate here if you only want to support a subset of the standard library in the beginning. But putting whole module behind feature gates that have yet to be introduced instead of stubbing them could require some work.

ivmarkov commented:
> Note that as part of only supporting fallible allocation in the linux kernel ([rust-lang/rfcs#3140](https://github.com/rust-lang/rfcs/pull/3140)) there is ongoing discussion around making std features optional and how to support that on the cargo side. [rust-lang/rfcs#3146](https://github.com/rust-lang/rfcs/pull/3146)
> 
> I'm not sure what exactly the roadmap on those parts are but it sounds like it would be appropriate here if you only want to support a subset of the standard library in the beginning. But putting whole modules behind feature gates that have yet to be introduced instead of stubbing them could require some work.

That's very useful info, thanks for sharing!

My problem with it is a little bit that this patch set has been in development since approx last November, and waiting for another Cargo nightly feature to first of all hit nightly, and then suffering through its stabilization pains, change of semantics and so on might be a bit too much. Especially considering that we are already sailing in an ocean of Cargo unstable features, that - when [changed](https://github.com/rust-lang/cargo/issues/9554#issuecomment-888095498) or regressed - [did hit us](https://github.com/rust-lang/cargo/issues/9641) in the [past](https://github.com/rust-lang/cargo/issues/8945):
- [build-std](https://github.com/rust-lang/wg-cargo-std-aware)
- [Extra link arg](https://github.com/rust-lang/cargo/issues/9426)
- [Configurable env](https://github.com/rust-lang/cargo/issues/9539)
- [Artifact dependencies](https://github.com/rust-lang/cargo/issues/9096) - lower prio; waiting on that, hasn't even hit nightly yet

So I would rather take a conservative approach here and stub what is not supported, if there is no strong, well-grounded objection from the Rust maintainers.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3213946223) [(plain)](https://github.com/rust-lang/rust/commit/772ac63208a5f6f6a359cc24fd6dd0b0d06e9bbb/checks/3213946223/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/os/espidf/raw.rs at line 9:
 )]
 
 
-use crate::os::raw::{c_long};
+use crate::os::raw::c_long;
 use crate::os::unix::raw::{gid_t, uid_t};
 
 #[stable(feature = "pthread_t", since = "1.8.0")]
Build completed unsuccessfully in 0:00:15
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/os/ios/raw.rs" "/checkout/library/std/src/os/espidf/mod.rs" "/checkout/library/std/src/os/espidf/raw.rs" "/checkout/library/std/src/os/ios/fs.rs" "/checkout/library/std/src/os/espidf/fs.rs" "/checkout/library/std/src/os/ios/mod.rs" "/checkout/library/std/src/sync/once/tests.rs" "/checkout/library/std/src/sync/rwlock.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3214022914) [(plain)](https://github.com/rust-lang/rust/commit/82aaf775409281566eb425c4e5b58d0bb66b0fa9/checks/3214022914/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
####                                                                       5.6%
######################################################################## 100.0%
extracting /checkout/obj/build/cache/2021-03-25/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
    Updating crates.io index
    Updating git repository `https://github.com/ivmarkov/libc.git`
---
  Downloaded miniz_oxide v0.4.0
  Downloaded hashbrown v0.11.0
   Compiling cc v1.0.69
    Checking core v0.0.0 (/checkout/library/core)
   Compiling libc v0.2.98 (https://github.com/ivmarkov/libc.git#b942bc82)
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.47
   Compiling unwind v0.0.0 (/checkout/library/unwind)
    Checking rustc-std-workspace-core v1.99.0 (/checkout/library/rustc-std-workspace-core)
---
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
Checking stage0 std artifacts (x86_64-unknown-linux-gnu -> i686-pc-windows-gnu)
   Compiling cc v1.0.69
    Checking core v0.0.0 (/checkout/library/core)
   Compiling libc v0.2.98 (https://github.com/ivmarkov/libc.git#b942bc82)
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.47
   Compiling unwind v0.0.0 (/checkout/library/unwind)
    Checking rustc-std-workspace-core v1.99.0 (/checkout/library/rustc-std-workspace-core)
---
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling cc v1.0.69
   Compiling core v0.0.0 (/checkout/library/core)
   Compiling libc v0.2.98 (https://github.com/ivmarkov/libc.git#b942bc82)
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.47
   Compiling unwind v0.0.0 (/checkout/library/unwind)
   Compiling rustc-std-workspace-core v1.99.0 (/checkout/library/rustc-std-workspace-core)
---
   Compiling regex v1.4.6
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 11.81s
tidy check
tidy error: invalid source: "git+https://github.com/ivmarkov/libc.git#b942bc82ac3b4104e6ce7284ceffeba8fbdc84b3"
Checking which error codes lack tests...
* 625 error codes
* highest error code: E0783
Found 499 error codes
Found 499 error codes
Found 0 error codes with no tests
Done!
* 345 features
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:14
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #87535) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Amanieu commented:
> Since support for atomics is a precondition for libStd, the `riscv32imc-esp-espidf` target additionally is configured in such a way, so as to emit libcalls to the `__sync*` & `__atomic*` GCC functions, which are already implemented in the ESP-IDF framework. If this modification is not acceptable, we can also live with only the `riscv32imac-esp-espidf` target as well. While the RiscV chips of Espressif lack native atomics support, the relevant instructions are transparently emulated in the ESP-IDF framework using invalid instruction trap. This modification was implemented specifically with Rust support in mind.

This is fine as long as the underlying implementation is lock-free (doesn't use a lock internally to emulate atomics). The requirement here is that a signal/interrupt handler should never deadlock if it tries to use an `Atomic*` which the interrupted thread was in the middle of accessing.

Generally this is implemented by disabling interrupt handlers or by having the interrupt handler restart the atomic sequence in the interrupted thread upon returning. We already have something similar for ARMv4 on Linux (atomic instructions were introduced in ARMv6) where support for atomic operations is provided by the kernel: https://github.com/rust-lang/compiler-builtins/blob/master/src/arm_linux.rs

ivmarkov commented:
> > Since support for atomics is a precondition for libStd, the `riscv32imc-esp-espidf` target additionally is configured in such a way, so as to emit libcalls to the `__sync*` & `__atomic*` GCC functions, which are already implemented in the ESP-IDF framework. If this modification is not acceptable, we can also live with only the `riscv32imac-esp-espidf` target as well. While the RiscV chips of Espressif lack native atomics support, the relevant instructions are transparently emulated in the ESP-IDF framework using invalid instruction trap. This modification was implemented specifically with Rust support in mind.
> 
> This is fine as long as the underlying implementation is lock-free (doesn't use a lock internally to emulate atomics). The requirement here is that a signal/interrupt handler should never deadlock if it tries to use an `Atomic*` which the interrupted thread was in the middle of accessing.

I think this is implemented by disabling the interrupt handlers upon entry in the `__sync` / `__atomic` functions and then restoring the previous state upon exit. In any case, @igrr can shed some light on both approaches.

> Generally this is implemented by disabling interrupt handlers or by having the interrupt handler restart the atomic sequence in the interrupted thread upon returning. We already have something similar for ARMv4 on Linux (atomic instructions were introduced in ARMv6) where support for atomic operations is provided by the kernel: https://github.com/rust-lang/compiler-builtins/blob/master/src/arm_linux.rs

Thanks for the link!

igrr commented:
> > > __sync* & __atomic* GCC functions, which are already implemented in the ESP-IDF framework

> > This is fine as long as the underlying implementation is lock-free (doesn't use a lock internally to emulate atomics).

> I think this is implemented by disabling the interrupt handlers upon entry in the __sync / __atomic functions and then restoring the previous state upon exit.

That's correct — `__sync_*` and `__atomic_*` built-ins are implement in esp-idf by disabling interrupts, performing the requested operation, then re-enabling interrupts. So the built-in implementation isn't expected to be preempted.

Currently this implementation is hosted in esp-idf but we could move (or copy) it into `rust-lang/compiler-builtins` at some point, if we wanted to support atomics on rv32imc without a dependency on esp-idf.

kennytm commented:
r? @Amanieu

Amanieu commented on library/std/src/sys/unix/os.rs:
```diff
@@ -566,10 +583,16 @@ pub fn unsetenv(n: &OsStr) -> io::Result<()> {
 
+#[cfg(target_os = "espidf")]
+pub fn page_size() -> usize {
+    32
```
What is the significance of this number? AFAIK RISC-V uses 4K pages. A comment here would be useful.

MabezDev replied:
Using the RISC-V standard of 4K pages might not make sense when add support for the Xtensa based Espressif chips (once support is upstreamed in LLVM of course). 

I also believe setting a smaller page size reduces the stack usage of threads where the stack size is not a multiple of the page size, according to the `Thread::new` impl: https://github.com/rust-lang/rust/blob/6fe0886723c9e08b800c9951f1c6f6a57b2bf22c/library/std/src/sys/unix/thread.rs#L63-L66

So I think we want to keep this number low.

igrr replied:
Since none of the chips supported by esp-idf so far has an MMU, we can indeed keep this number low. Perhaps 4 bytes would work?

However it is odd that this chip-specific property ends up depending on `#[cfg(target_os = "espidf")]`.

Long term, my preference would be implementing `sysconf`  and `_SC_PAGESIZE` in ESP-IDF and deciding on the right value there (possibly depending on the chip target).

Amanieu replied:
If this is just about the stack alignment then maybe it would be better to create a `stack_alignment` function in the thread module which can return `page_size` by default but a different value for `espidf`.

Reporting a page size of 32 is just confusing, even if this function is internal to `std`.

ivmarkov replied:
Please see [this summary of my latest changes](https://github.com/rust-lang/rust/pull/87666#issuecomment-893460598). No more `page_size` on ESP-IDF.

Amanieu reviewed (commented):
Overall, I don't feel very confident about the file descriptor duplication hack. It seems likely that it will break horribly with conversions to/from raw FDs.

I would feel more comfortable if `duplicate` just returned an error.

ivmarkov commented on library/std/src/sys/unix/fd.rs:
```diff
@@ -264,13 +288,61 @@ impl FileDesc {
+    }
+
+    fn remove_ref(&self) -> bool {
+        #[cfg(target_os = "espidf")]
+        {
+            let mut map = MAP.lock().unwrap();
+            match map.get_mut(&self.fd) {
+                Some(refcnt) => {
+                    *refcnt = *refcnt - 1;
+                    let close = *refcnt == 1;
+                    if close {
+                        map.remove(&self.fd);
+                    }
+                    close
+                }
+                None => true,
+            }
+        }
+
+        #[cfg(not(target_os = "espidf"))]
+        true
```
OK. Since there had been multiple concerns raised with the `duplicate` hack, I'll remove it. Thanks for the code cleanup though.

ivmarkov replied:
As per [my last set of changes](https://github.com/rust-lang/rust/pull/87666#issuecomment-893460598), the "duplicate" hack is no longer a part of the PR.

ivmarkov commented:
All,

I've pushed a small set of changes addressing the feedback accumulated so far. The changes are not force-pushed sop you can also only look at the latest (3rd) commit in the PR:

`compiler/rustc_target/src/spec/riscv32imc_esp_espidf.rs`
- I've adjusted the comment a bit as I would expect when/if this PR is merged to only keep this target, and remove the `riscv32imac-esp-espidf` one. We don't really need two targets for the same thing, and if you agree that emulating atomics with `__atomic*` /` __sync*` in ESP-IDF is the way forward, I'll remove the other target

`library/std/src/sys/unix/fd.rs`:
- I've removed the "duplicate" hack completely
- The `set_cloexec()` no-op is still necessary though, or else even basic TcpStream::connect & read test fails on the ESP-IDF
- I've changed a bit the existing `duplicate()` function, so that ESP-IDF gets `F_DUPFD` instead of `F_DUPFD_CLOEXEC`. Unlike `F_DUPFD_CLOEXEC`, `F_DUPFD` might be eventually supported (currently returns ENOSYS as expected).

`library/std/src/sys/unix/os.rs`:
- `page_size()` is now NOT defined on the ESP-IDF. That's OK, because:
(a) ESP-IDF does its own internal alignment of the stack size, so the Rust one can simply be switched off, as I did
(b) `thread::guard` and `stack_verflow.rs` where `page_size()` was also called are actually compiled as no-op on the ESP-IDF so we are in the end OK with NOT having `page_size` defined for the ESP-IDF at all. :)

`library/std/src/sys/unix/thread.rs`:
- I've implemented a code path for ESP-IDF where the stack-alignment logic is completely switched off. As per above, this logic is not necessary for the ESP-IDF, as it is anyway automatically aligning the stack size as per the requirements of the FreeRTOS.

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3252722681) [(plain)](https://github.com/rust-lang/rust/commit/a87168a6188a1decdc4a537ecbbcec4607f69c9b/checks/3252722681/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
From https://github.com/rust-lang/miri
 * branch              062cf07d98822beb4c5b1132f47b4397275f403a -> FETCH_HEAD
 * branch              06758c48bd7a77bb5aa43fc50cf344540ba5afef -> FETCH_HEAD
 * branch              dd7f545a69e4b720407e458bf4ade0b207bbf9ee -> FETCH_HEAD
fatal: remote error: upload-pack: not our ref 9ad6e5b8f68ee4bcd85886e50b8b0a70cbb91a52
Errors during submodule fetch:
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

ivmarkov commented on library/std/src/sys/unix/fd.rs:
```diff
@@ -264,13 +288,61 @@ impl FileDesc {
+            None => {
+                map.insert(self.fd, 2);
+            }
+        };
```
Indeed - but no longer relevant, as all of this code is removed.

ivmarkov commented:
Btw, any idea why these checks are now failing? Very strange, all does compile just fine locally for me...

Amanieu commented:
Github actions are on fire: https://www.githubstatus.com/

ivmarkov commented:
Github actions are back and the build is clean again.

Anything else that we should discuss regarding this PR? Is it in a mostly acceptable state for merging now?
* E.g. shall I now remove the `riscv32imac-esp-espidf` target and keep only the `riscv32imc-esp-espidf` one? 
* Anything else that needs cleanup?

Sorry - don't mean to push - just trying to make the best use of the few remaining days of my vacation.

Amanieu commented:
Everything looks good and ready to merge.

Regarding `riscv32imac-esp-espidf`, it's up to you. If `riscv32imc-esp-espidf` is sufficient and you don't have any actual (or future) hardware that supports atomics then there isn't much point adding it.

ivmarkov commented:
> Everything looks good and ready to merge.

That's great! Thank you for helping me drive this effort.

> Regarding `riscv32imac-esp-espidf`, it's up to you. If `riscv32imc-esp-espidf` is sufficient and you don't have any actual (or future) hardware that supports atomics then there isn't much point adding it.

I've removed the `riscv32imac-esp-espidf` target in favour of the `riscv32imc-esp-espidf` one which does generate `__async*` libcalls.

Let me know if you need me to rebase the whole PR as a single commit prior to merging.

One last thing: once the new version of the [libc crate is released](https://github.com/rust-lang/libc/pull/2312#issuecomment-891561270), we'll have to raise [the libc dependency in libStd from V0.2.98, to 0.2.99](https://github.com/rust-lang/rust/blob/master/library/std/Cargo.toml#L18), as this PR of course relies on the now-merged ESP-IDF changes in the libc crate.

But I guess we can do this post-factum once the new version of libc is released.

Disasm commented on library/std/src/sys/common/alloc.rs:
```diff
@@ -14,7 +14,8 @@ use crate::ptr;
     target_arch = "hexagon",
-    target_arch = "riscv32"
+    target_arch = "riscv32",
+    target_arch = "xtensa"
```
I think xtensa is not supported yet.

ivmarkov replied:
Support for Xtensa touches multiple crates and does not necessarily land into all of them in an atomic fashion:
- (1) `libc` crate: Here, it landed for the first time [> 1 year ago](https://github.com/rust-lang/libc/pull/1748) and was [recently extended](https://github.com/rust-lang/libc/pull/2310) with extra APIs and reworked to not really mean "Xtensa" but ESP-IDF for Risv32 *and* Xtensa and any architecture supported by ESP-IDf in fact.
- (2) `library/std`: Here, it lands now. The only place that needs "Xtensa" explicitly listed (as opposed to branching based on `target_os = "espidf"`) is this one AFAIR.
- (3) `rust/compiler/rustc_target`: Not landed yet, but we have a patchset for that. However, the Xtensa arch needs to be upstreamed in LLVM first.

Now, if you or somebody else feels uncomfortable having this conditional before the actual support for the Xtensa *targets* being upstreamed first, I can remove that line.

But if it does not do harm, why bother? We'll have to add it back anyway once (3) lands. And in the meantime, we'll know that libStd is Xtensa "ready" once the Xtensa targets get upstreamed.

Disasm commented on library/std/src/sys/unix/condvar.rs:
```diff
@@ -34,12 +34,23 @@ impl Condvar {
     ))]
     pub unsafe fn init(&mut self) {}
 
+    // NOTE: ESP-IDF's PTHREAD_COND_INITIALIZER support is not released yet
```
Maybe "is not implemented yet"?

Amanieu commented:
Let's wait for libc to publish a new release before merging this.

ivmarkov commented:
> Let's wait for libc to publish a new release before merging this.

No problem. Keeping fingers crossed that this would still happen before 1.56 is released.

ivmarkov commented:
Now that libc `0.2.99` is released, I've forced pushed the following two updates:
- Modification of `./Cargo.lock` and `./library/std/Cargo.toml` - only change is that the `libc` dependency is now raised to 0.2.99
- I've rebased all commits so far into a single one to keep the rust master branch history cleaner. Other than the rebase, no changes to the already reviewed code

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 459eaa6baea4127b37769d0b7944fa00c175e770 has been approved by `Amanieu`

<!-- @bors r=Amanieu 459eaa6baea4127b37769d0b7944fa00c175e770 -->
<!-- homu: {"type":"Approved","sha":"459eaa6baea4127b37769d0b7944fa00c175e770","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit 459eaa6baea4127b37769d0b7944fa00c175e770 with merge 6bed1f0bc3cc50c10aab26d5f94b16a00776b8a5...
<!-- homu: {"type":"BuildStarted","head_sha":"459eaa6baea4127b37769d0b7944fa00c175e770","merge_sha":"6bed1f0bc3cc50c10aab26d5f94b16a00776b8a5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3312048658)
Approved by: Amanieu
Pushing 6bed1f0bc3cc50c10aab26d5f94b16a00776b8a5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Amanieu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3312048658"},"merge_sha":"6bed1f0bc3cc50c10aab26d5f94b16a00776b8a5"} -->
